api/label/set.go

```diff
@@ -0,0 +1,372 @@
+// appear no more than once.
+func (l *Set) ToSlice() []core.KeyValue {
+	iter := l.Iter()
+	return iter.ToSlice()
+}
+
+// Equivalent returns a value that may be used as a map key.  The
+// Distinct type guarantees that the result will equal the equivalent
+// Distinct value of any label set with the same elements as this,
+// where sets are made unique by choosing the last value in the input
+// for any given key.
+func (l *Set) Equivalent() Distinct {
+	if l == nil || !l.equivalent.Valid() {
+		return emptySet.equivalent
+	}
+	return l.equivalent
+}
+
+// Equals returns true if the argument set is equivalent to this set.
+func (l *Set) Equals(o *Set) bool {
+	return l.Equivalent() == o.Equivalent()
+}
+
+// Encoded returns a the encoded form of this set, according to
+// `encoder`.  The result will be cached in this `*Set`.
+func (l *Set) Encoded(encoder Encoder) string {
+	if l == nil || encoder == nil {
+		return ""
+	}
+
+	id := encoder.ID()
+	if !id.Valid() {
+		// Invalid IDs are not cached.
+		return encoder.Encode(l.Iter())
+	}
+
+	var lookup *string
+	l.lock.Lock()
+	for idx := 0; idx < maxConcurrentEncoders; idx++ {
+		if l.encoders[idx] == id {
+			lookup = &l.encoded[idx]
+			break
+		}
+	}
+	l.lock.Unlock()
+
+	if lookup != nil {
+		return *lookup
+	}
+
+	r := encoder.Encode(l.Iter())
+
+	l.lock.Lock()
+	defer l.lock.Unlock()
+
+	for idx := 0; idx < maxConcurrentEncoders; idx++ {
+		if l.encoders[idx] == id {
+			return l.encoded[idx]
+		}
+		if !l.encoders[idx].Valid() {
+			l.encoders[idx] = id
+			l.encoded[idx] = r
+			return r
+		}
+	}
+
+	// TODO: This is a performance cliff.  Find a way for this to
+	// generate a warning.
+	return r
+}
+
+// NewSet returns a new `*Set`.  Except for empty sets, this method
+// adds an additional allocation compared with a call to
+// `NewSetWithSortable`.  See that method for more details.
+func NewSet(kvs ...core.KeyValue) Set {
+	// Check for empty set.
+	if len(kvs) == 0 {
+		return emptySet
+	}
+
+	return NewSetWithSortable(kvs, new(Sortable))
+}
+
+// NewSetWithSortable modifies the input slice, ensuring the
+// following:
+//
+// - Last-value wins semantics
+// - Caller sees the reordering, but doesn't lose values
+// - Repeated call preserve last-value wins.
+//
+// Note that methods are defined `*Set`, although no allocation for
+// `Set` is required.  Callers may avoid memory allocations by:
+//
+// - allocating a `Sortable` for use as a temporary in this method
+// - allocating a `Set` for storing the return value of this
+//   constructor.
+func NewSetWithSortable(kvs []core.KeyValue, tmp *Sortable) Set {
+	// The requitements stated above require that the stable
+	// result be placed in the end of the input slice, while
+	// overwritten values are swapped to the beginning.
+
+	// Check for empty set.
+	if len(kvs) == 0 {
+		return emptySet
+	}
+
+	*tmp = kvs
+
+	// Sort and de-duplicate.  Note: this use of `tmp`
```

Comment:
Might be better to scope this comment to the sorting only and put the de-duplication near the `for` loop.

```suggestion
	// Stable sort so the following de-duplication can implement last-value-wins semantics.
	// Note: this use of `tmp`
```

Reply:
Done.